### PR TITLE
add CFNumber.h bindings to number.rs

### PR DESCRIPTION
**Pull request by oopsies49**
Omitted some constants from the header which did not have a value.

### CODE REVIEW

File: src/number.rs [hunk 1]
```diff
@@ -1,7 +1,48 @@
 // Exports of <CoreFoundation/CFNumber.h>
 
+use libc::c_void;
+
 use base::*;
 
+#[repr(C)]
+struct __CFBoolean;
+pub type CFBooleanRef = __CFBoolean;
```
Comment (dcuddeback) [inline src/number.rs]: Should this be `*const __CFBoolean`? `CFNumber.h` defines `CFBooleanRef` like this:

``` c
typedef const struct CF_BRIDGED_TYPE(NSNumber) __CFBoolean * CFBooleanRef;
```

File: src/number.rs [hunk 2]
```diff
@@ -1,7 +1,48 @@
...
+
+extern "C" {
+    pub fn CFBooleanGetTypeID() -> CFTypeID;
+    pub fn CFBooleanGetValue(boolean: CFBooleanRef) -> Boolean;
+}
+
+pub type CFNumberType = CFTypeID;
```
Comment (dcuddeback) [inline src/number.rs]: Where does `CFTypeID` come from? The header file I'm looking at defines `CFNumberType` as `CFIndex`:

``` c
typedef CF_ENUM(CFIndex, CFNumberType) {
```

File: src/number.rs [hunk 3]
```diff
@@ -1,7 +1,48 @@
...
+pub const kCFNumberMaxType: CFNumberType = 16;
+
+#[repr(C)]
+struct __CFNumber;
+pub type CFNumberRef = *const __CFNumber;
+
 extern "C" {
     pub fn CFNumberGetTypeID() -> CFTypeID;
+    pub fn CFNumberCreate(alloc: CFAllocatorRef, theType: CFNumberType, valuePtr: *const c_void) -> CFNumberRef;
+    pub fn CFNumberGetValue(number: CFNumberRef, theType: CFNumberType, valuePtr: *const c_void) -> CFNumberRef;
```
Comment (dcuddeback) [inline src/number.rs]: `valuePtr` should be `*mut c_void`:

``` c
CF_EXPORT
Boolean CFNumberGetValue(CFNumberRef number, CFNumberType theType, void *valuePtr);
```

File: src/number.rs [hunk 4]
```diff
@@ -1,7 +1,48 @@
...
+    pub fn CFNumberGetType(number: CFNumberRef) -> CFNumberType;
+    pub fn CFNumberGetByteSize(number: CFNumberRef) -> CFIndex;
+    pub fn CFNumberIsFloatType(number: CFNumberRef) -> Boolean;
+    pub fn CFComparisonResult(number: CFNumberRef, otherNumber: CFNumberRef, context: *const c_void) -> CFComparisonResult;
```
Comment (dcuddeback) [inline src/number.rs]: Is this function supposed to be `CFNumberCompare`?

``` c
CF_EXPORT
CFComparisonResult CFNumberCompare(CFNumberRef number, CFNumberRef otherNumber, void *context);
```

File: src/lib.rs [hunk 5]
```diff
@@ -17,3 +18,4 @@ pub mod data;
 pub mod dictionary;
 pub mod locale;
 pub mod string;
+pub mod number;
```
Comment (dcuddeback) [inline src/lib.rs:21]: I'd like to keep the modules alphabetized if possible.

**Comment by dcuddeback** [comment]
@oopsies49 Thanks for doing this. I found a couple discrepancies between the definitions in the PR and the header file that I get in Xcode. Just curious, which version of the header file are you using? I'm using the one that comes with the OS X 10.10 framework.

Do  you mind reordering the function in the second `extern "C"` block to match the order in the `CFNumber.h` header file? This is the order I see in the header file:

``` c
CFTypeID CFNumberGetTypeID(void);
CFNumberRef CFNumberCreate(CFAllocatorRef allocator, CFNumberType theType, const void *valuePtr);
CFNumberType CFNumberGetType(CFNumberRef number);
CFIndex CFNumberGetByteSize(CFNumberRef number);
Boolean CFNumberIsFloatType(CFNumberRef number);
Boolean CFNumberGetValue(CFNumberRef number, CFNumberType theType, void *valuePtr);
CFComparisonResult CFNumberCompare(CFNumberRef number, CFNumberRef otherNumber, void *context);
```

File: src/number.rs (hunk 2 above)
Comment (oopsies49) [inline src/number.rs]: The values of the enum would still be defined as `CFNumberType` though, correct?

``` rust
pub const kCFNumberSInt8Type: CFNumberType = 1;
```

**Comment by oopsies49** [comment]
Happy to contribute. I'm also on OSX 10.10... not sure how to get a version number of the framework. Based upon your suggested changes, I would say we are on the same version. Just some late night oversights to fix is all.

File: src/number.rs [hunk 6]
```diff
@@ -1,7 +1,47 @@
...
+pub const kCFNumberCGFloatType: CFNumberType = 16;
+pub const kCFNumberMaxType: CFNumberType = 16;
+
+#[repr(C)]
+struct __CFNumber;
+pub type CFNumberRef = *const __CFNumber;
+
 extern "C" {
     pub fn CFNumberGetTypeID() -> CFTypeID;
+    pub fn CFNumberCreate(alloc: CFAllocatorRef, theType: CFNumberType, valuePtr: *mut c_void) -> CFNumberRef;
```
Comment (dcuddeback) [inline src/number.rs]: Oops. Looks like this got changed from `*const` to `*mut` in your last commit. Should be `*const`.

File: src/number.rs [hunk 7]
```diff
@@ -1,7 +1,47 @@
...
+    pub fn CFNumberGetType(number: CFNumberRef) -> CFNumberType;
+    pub fn CFNumberGetByteSize(number: CFNumberRef) -> CFIndex;
+    pub fn CFNumberIsFloatType(number: CFNumberRef) -> Boolean;
+    pub fn CFNumberGetValue(number: CFNumberRef, theType: CFNumberType, valuePtr: *mut c_void) -> CFNumberRef;
```
Comment (dcuddeback) [inline src/number.rs]: I overlooked this yesterday. The return type for `CFNumberGetValue` is `Boolean`.

**Comment by oopsies49** [comment]
One thing I noticed when I was double checking things: I think `CFNumberCompare` parameter context should be `*mut`. Nothing in the header indicates that it is used for anything, it is not indicated to be `const`.

**Comment by dcuddeback** [comment]
@oopsies49 Yeah, you're right. Good catch. The documentation just says "Pass `NULL`," so I assume it's not used. I agree that it should be `*mut` to match the C prototype.

**Comment by dcuddeback** [comment]
@oopsies49 The documentation that I'm referring to is here: https://developer.apple.com/library/mac/documentation/CoreFoundation/Reference/CFNumberRef/#//apple_ref/c/func/CFNumberCompare

**Comment by dcuddeback** [comment]
@oopsies49 Looks good! :+1:

**Comment by dcuddeback** [comment]
@oopsies49 Published it to crates.io as v0.1.1.

**Pull request merged**